### PR TITLE
Extract only the target frame feature everytime.

### PR DESCRIPTION
**Pull request by wuyongfa-genius**
When evaluating on VOS datasets, we only need to extract the target frame feature everytime and store n_last_frame_features in the Queue. This will save a lot of computation and make the evaluation much faster.

### CODE REVIEW

**Comment by mathildecaron31** [comment]
Hi @wuyongfa-genius 

thanks a lot for this PR ! Have you tested the code ? That would be great to make sure that we reproduce the results for DINO with DeiT-S/16 architecture before merging.

**Comment by mathildecaron31** [comment]
I've tested quickly and it does not seem to reproduce the results:

--------------------------- Global results for val ---------------------------                                         
J&F-Mean   J-Mean  J-Recall   J-Decay   F-Mean  F-Recall   F-Decay                                                                                                                                                                            
 0.57754  0.56166  0.659446  0.165053  0.59342  0.709795  0.200766    

where J&F-Mean should be ~0.6177

I think I have spotted the bug. Please update the PR accordingly and I'll run a test again.

File: eval_video_segmentation.py [hunk 1]
```diff
@@ -107,37 +110,20 @@ def norm_mask(mask):
...
-        out = out.permute(0, 3, 1, 2)
-
-        if curr_i == 0:
-            feature_maps = out
-        else:
-            feature_maps = torch.cat((feature_maps, out))
-        curr_i += args.bs
-        del out  # free memory
+    ## we only need to extract feature of the target frame
+    feat_tar, h, w = extract_feature(model, frame_tar, return_h_w=True)
```
Comment (mathildecaron31) [inline eval_video_segmentation.py:118]: I see two options to fix the bug:

1) `feat_tar, w, h = extract_feature(model, frame_tar, return_h_w=True)`

OR

2) Replace line 133 `mask_neighborhood = restrict_neighborhood(w, h)` by `mask_neighborhood = restrict_neighborhood(h, w)`

Comment (mathildecaron31) [reply eval_video_segmentation.py:118]: I would recommend option 2.

**Review by mathildecaron31 — CHANGES_REQUESTED** [review]
Thanks for the PR ! Please address my comment since the code is not reproducing the results atm.

**Comment by wuyongfa-genius** [comment]
Sorry for the trouble... Have fixed the bugs about sizes and shapes and made some changes to function read_seg in order that the 'h' is corresponded to real height and 'w' real width...And also tested with deit_small16, the results seemed fine.
Per-sequence results saved in /data/un_or_self_supervised/dino/deit_small16_results/per-sequence_results-val.csv  
--------------------------- Global results for val ---------------------------
 J&F-Mean    J-Mean  J-Recall   J-Decay    F-Mean  F-Recall   F-Decay
 0.617763  0.601762  0.740277  0.129538  0.633764  0.763532  0.188162

**Comment by mathildecaron31** [comment]
Looks great ! Thanks a lot @wuyongfa-genius :) :)

**Pull request merged**